presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java

```diff
@@ -0,0 +1,482 @@
+import static com.facebook.presto.spi.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.sql.analyzer.ExpressionAnalyzer.getExpressionTypes;
+import static com.facebook.presto.sql.planner.SymbolsExtractor.extractAll;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.facebook.presto.sql.planner.plan.Patterns.project;
+import static com.facebook.presto.sql.planner.plan.Patterns.sort;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.facebook.presto.sql.planner.plan.Patterns.unnest;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+
+public class PushDownDereferences
+{
+    private final Metadata metadata;
+    private final SqlParser sqlParser;
+
+    public PushDownDereferences(Metadata metadata, SqlParser sqlParser)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+    }
+
+    public Set<Rule<?>> rules()
+    {
+        return ImmutableSet.of(
+                new PushDownDereferenceThroughJoin(metadata, sqlParser),
+                new PushDownDereferenceThroughSort(metadata, sqlParser),
+                new PushDownDereferenceThroughUnnest(metadata, sqlParser),
+                new PushDownDereferenceThroughProject(metadata, sqlParser));
+    }
+
+    private abstract class DereferencePushDownRule<N extends PlanNode>
+            implements Rule<ProjectNode>
+    {
+        private final Capture<N> targetCapture = newCapture();
+        private final Pattern<N> targetPattern;
+
+        protected final Metadata metadata;
+        protected final SqlParser sqlParser;
+
+        protected DereferencePushDownRule(Metadata metadata, SqlParser sqlParser, Pattern<N> targetPattern)
+        {
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+            this.targetPattern = requireNonNull(targetPattern, "targetPattern is null");
+        }
+
+        @Override
+        public Pattern<ProjectNode> getPattern()
+        {
+            return project().with(source().matching(targetPattern.capturedAs(targetCapture)));
+        }
+
+        @Override
+        public Result apply(ProjectNode node, Captures captures, Context context)
+        {
+            N child = captures.get(targetCapture);
+            Map<DereferenceExpression, Symbol> expressions = getDereferenceSymbolMap(node.getAssignments().getExpressions().stream().collect(toImmutableList()), context, metadata, sqlParser);
```

Comment:
`.stream().collect(toImmutableList())` is unnecessary; just change the type of the first argument of getDereferenceSymbolMap to `Collection<Expression>`.